client/app/modules/tests-sessions/access-key-modal/access-key-modal.controller.js

```diff
@@ -0,0 +1,447 @@
+            defaultItem = childControl.items[0];
+        }
+
+        return defaultItem;
+    }
+
+    //TODO: looks like we can get rid of this 'versions' configs because its functionality can be covered by common approach (variants)
+    function getDefaultVersionControl(childControl) {
+        let defaultItem;
+
+        if (!childControl.items.length) {
+            return  defaultItem;
+        }
+
+        //select by config's default value
+        if (!defaultItem && childControl.data['default-versions']) {
+            if (typeof childControl.data['default-versions'] === 'string') {
+                defaultItem = childControl.items.find(item => item.id === childControl.data['default-versions']);
+            } else {
+                defaultItem = childControl.items.find(item =>  childControl.data['default-versions'].includes(item.id));
+            }
+        }
+        //select first item in array
+        if (!defaultItem) {
+            defaultItem = childControl.items[0];
+        }
+
+        return defaultItem;
+    }
+
+    function filterPlatformModel() {
+        const keys = vm.platformControls.map(control => control.key);
+        const newModel = {};
+
+        newModel[vm.platformsConfig.rootKey] = vm.platformModel[vm.platformsConfig.rootKey];
+
+        vm.platformModel = keys.reduce((out, key) => {
+            out[key] = vm.platformModel[key];
+
+            return out;
+        }, newModel);
+    }
+
+    function updateEditorModel() {
+        $timeout(() => {
+            const selectedLanguage = getSelectedLanguage();
+
+            if (selectedLanguage && selectedLanguage.snippet) {
+                const data = Object.keys(vm.platformModel).reduce((acc, key) => {
+                    acc[key] = vm.platformModel[key].value;
+
+                    return acc;
+                }, {});
+
+                if (vm.accessSettings.ZEBRUNNER_USER) { data.username = vm.accessSettings.ZEBRUNNER_USER; }
+                if (vm.accessSettings.ZEBRUNNER_PASSWORD) { data.password = vm.accessSettings.ZEBRUNNER_PASSWORD; }
+
+                vm.aceModel = replacePlaceholders(selectedLanguage.snippet, data);
+                vm.lastUpdatedAceModel = vm.aceModel;
+            }
+        }, 0);
+    }
+
+    function replacePlaceholders(text, data) {
+        return text.replace(/\${([^{}]*)}/g, function (a, b) {
+            let replacer = data[`capabilities.${b}`] || data[b];
+
+            if (b === 'platformName') {
+                if (replacer === '*') {
+                    replacer = 'any';
+                }
+                replacer = replacer.toUpperCase();
+            }
+
+            return replacer && (typeof replacer === 'string' || typeof replacer === 'number') ? replacer : a;
+        });
+    }
+
+    function onCodeCopy() {
+        $timeout(() => {
+            vm.aceModel.copyToClipboard();
+            messageService.success('Code copied to clipboard');
+            $mdDialog.hide();
+        }, 0);
```

Comment:
maybe it's better to use the third argument `false`

Reply:
fixed